File: bench/inference_gpu.exs
# GPU inference latency benchmark for ExTorch.Export.
#
# Usage:  mix run bench/inference_gpu.exs
#
# Loads each popular_models fixture on :cuda, runs the saved input on :cuda
# through ExTorch.Export.forward N times, and reports min / median / mean
# latency. Requires a CUDA-capable libtorch build (check via
# `ExTorch.cuda_is_available/0`).

unless ExTorch.Native.cuda_is_available() do
  IO.puts("ERROR: CUDA is not available in this libtorch build.")
  IO.puts("Make sure priv/native/libtorch is a CUDA build (2.11.0+cuXYZ).")
  System.halt(1)
end

# Inference-mode fixes: disable autograd graph build-up process-wide.
# clear_cpu_affinity is harmless on GPU (useful work happens on the GPU,
# not on CPU threads) but costs nothing.
ExTorch.set_grad_enabled(false)
ExTorch.clear_cpu_affinity()

IO.puts("ExTorch backend: #{ExTorch.Native.aten_backend_info()}")
IO.puts("CUDA available:  true")
IO.puts("")

defmodule InferenceGpu do
  @fixtures Path.join([__DIR__, "..", "test", "fixtures"])
  @warmup 5
  @iters 30

  @models [
    {"alexnet",            {1, 3, 224, 224},  {1, 1000}},
    {"vgg11",              {1, 3, 224, 224},  {1, 1000}},
    {"squeezenet",         {1, 3, 224, 224},  {1, 1000}},
    {"mobilenetv2",        {1, 3, 224, 224},  {1, 1000}},
    {"resnet18",           {1, 3, 224, 224},  {1, 1000}},
    {"resnet50",           {1, 3, 224, 224},  {1, 1000}},
    {"simple_transformer", {1, 16, 32},       {1, 10}},
    {"mini_bert",          {1, 16, 32},       {1, 10}},
    {"autoencoder",        {1, 784},          {1, 784}},
    {"conv_autoencoder",   {1, 3, 32, 32},    {1, 3, 32, 32}},
    {"simple_lstm",        {1, 16, 32},       {1, 10}}
  ]

  def run do
    IO.puts("== ExTorch.Export GPU latency  (#{@iters} iters after #{@warmup} warmup) ==\n")
    IO.puts(:io_lib.format(~c"~-22s ~10s ~10s ~10s", [~c"model", ~c"min(ms)", ~c"median(ms)", ~c"mean(ms)"]))
    IO.puts(String.duplicate("-", 56))

    for {name, in_shape, _out_shape} <- @models do
      path = Path.join(@fixtures, "#{name}.pt2")
      bin = Path.join(@fixtures, "#{name}_input.bin")
      if File.exists?(path) and File.exists?(bin) do
        # Load model with weights placed on CUDA up front.
        model = ExTorch.Export.load(path, device: :cuda)
        # Load input on CPU then move to CUDA.
        cpu_input = ExTorch.Native.from_binary(File.read!(bin), in_shape, :float32)
        input = ExTorch.Tensor.to(cpu_input, device: :cuda)

        # Warmup — first CUDA calls compile/select kernels and allocate
        # the CUDA caching allocator pool; we need more warmups than CPU.
        for _ <- 1..@warmup, do: ExTorch.Export.forward(model, [input])
        ExTorch.cuda_synchronize()

        samples =
          for _ <- 1..@iters do
            # Sync before AND after so the measured interval only
            # contains this forward's kernels. CUDA launches are async.
            {us, _} = :timer.tc(fn ->
              _out = ExTorch.Export.forward(model, [input])
              ExTorch.cuda_synchronize()
            end)
            us / 1000.0
          end

        sorted = Enum.sort(samples)
        min_v = List.first(sorted)
        median = Enum.at(sorted, div(length(sorted), 2))
        mean = Enum.sum(samples) / length(samples)
        IO.puts(:io_lib.format(~c"~-22s ~10.2f ~10.2f ~10.2f", [String.to_charlist(name), min_v, median, mean]))
      else
        IO.puts(:io_lib.format(~c"~-22s    SKIP (fixture missing)", [String.to_charlist(name)]))
      end
    end
  end
end

InferenceGpu.run()

File: bench/inference_gpu.py
"""GPU inference latency benchmark for PyTorch reference.

Usage:  .venv/bin/python bench/inference_gpu.py

Loads each .pt2 fixture on CUDA, runs the saved input on CUDA N times,
and reports min / median / mean latency. Requires a CUDA-capable PyTorch
build (check via `torch.cuda.is_available()`).

Paired with bench/inference_gpu.exs to measure ExTorch vs PyTorch on the
same GPU.
"""

import os
import statistics
import sys
import time

import torch

if not torch.cuda.is_available():
    print("ERROR: torch.cuda.is_available() == False")
    sys.exit(1)

# Match ExTorch's inference-mode setup: disable gradient tracking globally.
torch.set_grad_enabled(False)

DEVICE = torch.device("cuda")
FIXTURES_DIR = os.path.join(os.path.dirname(__file__), "..", "test", "fixtures")
WARMUP = 5
ITERS = 30

MODELS = [
    ("alexnet",            (1, 3, 224, 224),  (1, 1000)),
    ("vgg11",              (1, 3, 224, 224),  (1, 1000)),
    ("squeezenet",         (1, 3, 224, 224),  (1, 1000)),
    ("mobilenetv2",        (1, 3, 224, 224),  (1, 1000)),
    ("resnet18",           (1, 3, 224, 224),  (1, 1000)),
    ("resnet50",           (1, 3, 224, 224),  (1, 1000)),
    ("simple_transformer", (1, 16, 32),       (1, 10)),
    ("mini_bert",          (1, 16, 32),       (1, 10)),
    ("autoencoder",        (1, 784),          (1, 784)),
    ("conv_autoencoder",   (1, 3, 32, 32),    (1, 3, 32, 32)),
    ("simple_lstm",        (1, 16, 32),       (1, 10)),
]


def time_one(runner, x):
    # Use CUDA events for accurate GPU timing: launch + synchronize.
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    runner(x)
    end.record()
    torch.cuda.synchronize()
    return start.elapsed_time(end)  # ms


def main():
    print(f"Device: {torch.cuda.get_device_name(0)}")
    print(f"torch:  {torch.__version__}")
    print(f"CUDA:   {torch.version.cuda}")
    print(f"cuDNN:  {torch.backends.cudnn.version()}  enabled={torch.backends.cudnn.enabled}")
    print()
    print(f"== PyTorch torch.export.load GPU latency  ({ITERS} iters after {WARMUP} warmup) ==\n")
    print(f"{'model':<22} {'min(ms)':>10} {'median(ms)':>10} {'mean(ms)':>10}")
    print("-" * 56)

    for name, in_shape, _out_shape in MODELS:
        pt2 = os.path.join(FIXTURES_DIR, f"{name}.pt2")
        bin_path = os.path.join(FIXTURES_DIR, f"{name}_input.bin")
        if not os.path.exists(pt2) or not os.path.exists(bin_path):
            print(f"{name:<22}    SKIP (fixture missing)")
            continue

        with open(bin_path, "rb") as f:
            raw = f.read()
        cpu_input = torch.frombuffer(bytearray(raw), dtype=torch.float32).reshape(in_shape)
        input_tensor = cpu_input.to(DEVICE)

        ep = torch.export.load(pt2)
        runner = ep.module().to(DEVICE)

        # Warmup: compile kernel selection / allocate CUDA caches
        for _ in range(WARMUP):
            runner(input_tensor)
        torch.cuda.synchronize()

        samples = [time_one(runner, input_tensor) for _ in range(ITERS)]
        samples_sorted = sorted(samples)
        print(
            f"{name:<22} "
            f"{samples_sorted[0]:>10.2f} "
            f"{samples_sorted[len(samples_sorted) // 2]:>10.2f} "
            f"{statistics.mean(samples):>10.2f}"
        )


if __name__ == "__main__":
    main()
